compiler.js

```diff
@@ -165,10 +237,128 @@ class SlangCompiler
         return result;
     }
 
+    // If user entrypoint name imageMain or printMain, we will load the pre-built main modules because they
+    // are defined in those modules. Otherwise, we will only need to load the user module and find the entry
+    // point in the user module.
+    shouldLoadMainModule(entryPointName)
+    {
+        return entryPointName == "imageMain" || entryPointName == "printMain";
+    }
+
+    // Since we will not let user to change the entry point code, we can precompile the entry point module
+    // and reuse it for every compilation.
+
+    compileEntryPointModule(slangSession, moduleName)
+    {
+        var module = slangSession.loadModuleFromSource(this.mainModules[moduleName].source, moduleName, '/' + moduleName + '.slang');
+
+        if (!module) {
+            var error = this.slangWasmModule.getLastError();
+            console.error(error.type + " error: " + error.message);
+            this.diagnosticsMsg+=(error.type + " error: " + error.message);
+            return null;
+        }
+
+        // we use the same entry point name as module name
+        var entryPoint = this.findEntryPoint(module, moduleName, SlangCompiler.SLANG_STAGE_COMPUTE);
+        if (!entryPoint)
+            return null;
+
+        return {module: module, entryPoint: entryPoint};
+
+    }
+
+    getPrecompiledProgram(slangSession, moduleName)
+    {
+        if (moduleName != "printMain" && moduleName != "imageMain")
+            return null;
+
+        if (this.mainModules[moduleName].module)
+            return this.mainModules[moduleName];
+
+        this.mainModules[moduleName] = this.compileEntryPointModule(slangSession, moduleName);
+
+        this.shaderType = SlangCompiler.RENDER_SHADER;
+        return this.mainModules[moduleName];
+    }
+
+    addActiveEntryPoints(slangSession, shaderSource, entryPointName, isWholeProgram, userModule, componentList)
+    {
+        if (entryPointName == "" && !isWholeProgram)
+        {
+            this.diagnosticsMsg+=("error: No entry point specified");
+            return false;
+        }
+
+        // For now, we just don't allow user to define imageMain or printMain as entry point name for simplicity
+        var count = userModule.getDefinedEntryPointCount();
+        for (var i = 0; i < count; i++)
+        {
+            var entryPointName = userModule.getDefinedEntryPoint(i).getName();
```

Comment:
this will leak the entrypoint returned by `getDefinedEntryPoint()`.

Reply:
does that matter? `findDefinedEntryPoints` also leak the entrypoint.

Reply:
I deleted the entrypoints there:
```
            for (var i = 0; i < count; i++)
            {
                var entryPoint = module.getDefinedEntryPoint(i);
                result.push(entryPoint.getName());
                entryPoint.delete();
            }
```
Let's aim to not leak at least for frequently called operations.